mixer/pkg/config/crd/admit.go

```diff
@@ -0,0 +1,387 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crd
+
+// Note: this file is copied from pilot/platform/kube/admit. Ideally we should
+// make a common component for this server side part, accepting component-specific
+// validation logic.
+// TODO: do so.
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	"k8s.io/api/admission/v1alpha1"
+	admissionregistrationv1alpha1 "k8s.io/api/admissionregistration/v1alpha1"
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/kubernetes"
+	admissionClient "k8s.io/client-go/kubernetes/typed/admissionregistration/v1alpha1"
+	"k8s.io/client-go/tools/cache"
+
+	"istio.io/istio/mixer/pkg/config/store"
+)
+
+const (
+	secretServerKey  = "server-key.pem"
+	secretServerCert = "server-cert.pem"
+	secretCACert     = "ca-cert.pem"
+)
+
+// ControllerOptions contains the configuration for the Istio Pilot validation
+// admission controller.
+type ControllerOptions struct {
+	// The name of the resources this controller can admit.
+	ResourceNames []string
+
+	// ExternalAdmissionWebhookName is the name of the
+	// ExternalAdmissionHook which describes he external admission
+	// webhook and resources and operations it applies to.
+	ExternalAdmissionWebhookName string
+
+	// ServiceName is the service name of the webhook.
+	ServiceName string
+
+	// ServiceNamespace is the namespace of the webhook service.
+	ServiceNamespace string
+
+	// ValidateNamespaces is a list of names to validate. Any
+	// namespace not in this list is unconditionally validated as
+	// good. This is useful when multiple validators are running in
+	// the same cluster managing different sets of namespaces
+	// (e.g. shared test clusters).
```

Comment:
This option is very test-infra specific. If it is not meant for prod use, we should make it clear in the comments.

Reply:
added here, and the flag description.